File: plugins/plugin-imessage/rust/src/providers/chat_context.rs
//! Chat context provider for iMessage

use crate::service::IMessageService;
use crate::IMESSAGE_SERVICE_NAME;
use async_trait::async_trait;
use elizaos::{IAgentRuntime, Memory, Provider, State};
use tracing::debug;

/// Provider that supplies iMessage chat context
pub struct ChatContextProvider;

impl ChatContextProvider {
    /// Creates a new chat context provider
    pub fn new() -> Self {
        Self
    }
}

impl Default for ChatContextProvider {
    fn default() -> Self {
        Self::new()
    }
}

#[async_trait]
impl Provider for ChatContextProvider {
    fn name(&self) -> &str {
        "IMESSAGE_CHAT_CONTEXT"
    }

    fn description(&self) -> &str {
        "Provides information about the current iMessage chat context"
    }

    async fn get(
        &self,
        runtime: &dyn IAgentRuntime,
        message: &Memory,
        _state: Option<&State>,
    ) -> String {
        let service = match runtime.get_service::<IMessageService>(IMESSAGE_SERVICE_NAME) {
            Some(s) => s,
            None => return String::new(),
        };

        let room = match runtime.get_room(&message.room_id).await {
            Ok(Some(r)) => r,
            _ => return String::new(),
        };

        // Only provide context for iMessage channels
        if room.source.as_deref() != Some("imessage") {
            return String::new();
        }

        let channel_id = match room.channel_id {
            Some(ref id) => id,
            None => return String::new(),
        };

        format_chat_context(channel_id, room.name.as_deref())
    }
}

fn format_chat_context(channel_id: &str, name: Option<&str>) -> String {
    let mut lines = vec![
        "# iMessage Chat Context".to_string(),
        String::new(),
    ];

    if let Some(chat_name) = name {
        lines.push(format!("- Chat: {}", chat_name));
    }

    lines.push(format!("- Channel ID: {}", channel_id));
    lines.push(String::new());
    lines.push(
        "Note: This conversation is happening through iMessage. Be conversational and friendly."
            .to_string(),
    );

    lines.join("\n")
}

File: plugins/plugin-imessage/src/providers/chat-context.ts
import type { IAgentRuntime, Memory, Provider, ProviderResult, State } from "@elizaos/core";

function firstString(...values: unknown[]): string | undefined {
  for (const value of values) {
    if (typeof value === "string" && value.trim()) return value.trim();
  }
  return undefined;
}

export const chatContextProvider: Provider = {
  name: "imessageChatContext",
  description: "Provides iMessage chat context for the current conversation.",
  descriptionCompressed: "Current iMessage chat handle, chat ID, type, and display name.",
  dynamic: true,
  contextGate: { anyOf: ["phone", "social", "connectors"] },
  cacheStable: false,
  cacheScope: "turn",
  contexts: ["phone", "social", "connectors"],

  get: async (_runtime: IAgentRuntime, message: Memory, state?: State): Promise<ProviderResult> => {
    const stateData = (state?.data ?? {}) as Record<string, unknown>;
    const metadata = (message.metadata ?? {}) as Record<string, unknown>;
    const imessageMetadata =
      metadata.imessage && typeof metadata.imessage === "object"
        ? (metadata.imessage as Record<string, unknown>)
        : {};

    const chatId = firstString(stateData.chatId, imessageMetadata.chatId);
    const handle = firstString(
      stateData.handle,
      metadata.imessageHandle,
      imessageMetadata.userId,
      imessageMetadata.id
    );
    const chatType = firstString(stateData.chatType, metadata.chatType);
    const displayName = firstString(
      stateData.displayName,
      metadata.imessageContactName,
      metadata.entityName,
      imessageMetadata.name
    );

    if (!chatId && !handle && !chatType && !displayName) {
      return {
        data: { available: false },
        values: {},
        text: "",
      };
    }

    const lines = [
      "Current iMessage chat context:",
      chatId ? `- Chat ID: ${chatId}` : null,
      handle ? `- Handle: ${handle}` : null,
      chatType ? `- Chat type: ${chatType}` : null,
      displayName ? `- Display name: ${displayName}` : null,
    ].filter((line): line is string => Boolean(line));

    return {
      data: {
        available: true,
        chatId,
        handle,
        chatType,
        displayName,
      },
      values: {
        chatId,
        handle,
        chatType,
        displayName,
      },
      text: lines.join("\n"),
    };
  },
};
